### PR TITLE
Replace cudf's concurrent_ordered_map with cuco::static_map in semi/anti joins

### LINKED ISSUE #9586: [FEA] Refactor semi/anti join to use cuCollections
**Is your feature request related to a problem? Please describe.**

https://github.com/rapidsai/cudf/pull/8934 refactored inner/outer/full joins to use the cuCollections multi-map, but not semi or anti-join. 

**Describe the solution you'd like**

Refactor semi/anti-join to use `cuco::static_map`. 

**Additional context**

This is a precondition of support semi-joins with equality + conditional as described in https://github.com/rapidsai/cudf/issues/5401 

There were a number of improvements we made to `cuco::static_multimap` that haven't been ported to `static_map` yet that we may want to do alongside this effort.

### PR DESCRIPTION
**Pull request by vyasr**
This PR resolves #9586, replacing the hash table used in semi and anti joins with cuco::static_map. It depends on NVIDIA/cuCollections#118. At present the code is slower than the original version, so we'll probably want to make some optimizations in cuco before merging this.

### CODE REVIEW

**Comment by vyasr** [comment]
Current benchmarks:

<details><summary>Benchmarks</summary>
<p>
Old code:

```
Join<int32_t, int32_t>/left_semi_join_32bit/100000/100000/manual_time                                    0.156 ms        0.177 ms         4409
Join<int32_t, int32_t>/left_semi_join_32bit/100000/400000/manual_time                                    0.259 ms        0.278 ms         2699
Join<int32_t, int32_t>/left_semi_join_32bit/100000/1000000/manual_time                                   0.474 ms        0.493 ms         1479
Join<int32_t, int32_t>/left_semi_join_32bit/10000000/10000000/manual_time                                 30.8 ms         30.9 ms           23
Join<int32_t, int32_t>/left_semi_join_32bit/10000000/40000000/manual_time                                 61.3 ms         61.3 ms           11
Join<int32_t, int32_t>/left_semi_join_32bit/10000000/100000000/manual_time                                 122 ms          122 ms            6
Join<int32_t, int32_t>/left_semi_join_32bit/100000000/100000000/manual_time                                313 ms          313 ms            2
Join<int32_t, int32_t>/left_semi_join_32bit/80000000/240000000/manual_time                                 414 ms          414 ms            2
Join<int64_t, int64_t>/left_semi_join_64bit/50000000/50000000/manual_time                                  164 ms          164 ms            4
Join<int64_t, int64_t>/left_semi_join_64bit/40000000/120000000/manual_time                                 225 ms          225 ms            3
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/100000/manual_time                              0.151 ms        0.173 ms         4556
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/400000/manual_time                              0.197 ms        0.218 ms         3482
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/1000000/manual_time                             0.265 ms        0.283 ms         2654
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/10000000/10000000/manual_time                           7.08 ms         7.10 ms           99
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/10000000/40000000/manual_time                           14.6 ms         14.6 ms           48
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/10000000/100000000/manual_time                          32.3 ms         32.3 ms           18
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000000/100000000/manual_time                         76.7 ms         76.7 ms            8
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/80000000/240000000/manual_time                           107 ms          107 ms            5
Join<int64_t, int64_t>/left_semi_join_64bit_nulls/50000000/50000000/manual_time                           43.5 ms         43.6 ms           13
Join<int64_t, int64_t>/left_semi_join_64bit_nulls/40000000/120000000/manual_time                          66.0 ms         66.0 ms            8
```

New code:
```
Join<int32_t, int32_t>/left_semi_join_32bit/100000/100000/manual_time                                    0.426 ms        0.446 ms         1613
Join<int32_t, int32_t>/left_semi_join_32bit/100000/400000/manual_time                                    0.840 ms        0.860 ms          833
Join<int32_t, int32_t>/left_semi_join_32bit/100000/1000000/manual_time                                    1.44 ms         1.45 ms          490
Join<int32_t, int32_t>/left_semi_join_32bit/10000000/10000000/manual_time                                 32.2 ms         32.2 ms           22
Join<int32_t, int32_t>/left_semi_join_32bit/10000000/40000000/manual_time                                 77.0 ms         77.0 ms            9
Join<int32_t, int32_t>/left_semi_join_32bit/10000000/100000000/manual_time                                 167 ms          167 ms            4
Join<int32_t, int32_t>/left_semi_join_32bit/100000000/100000000/manual_time                                324 ms          324 ms            2
Join<int32_t, int32_t>/left_semi_join_32bit/80000000/240000000/manual_time                                 505 ms          505 ms            2
Join<int64_t, int64_t>/left_semi_join_64bit/50000000/50000000/manual_time                                  194 ms          194 ms            4
Join<int64_t, int64_t>/left_semi_join_64bit/40000000/120000000/manual_time                                 322 ms          322 ms            2
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/100000/manual_time                              0.191 ms        0.212 ms         3720
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/400000/manual_time                              0.229 ms        0.250 ms         3001
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/1000000/manual_time                             0.359 ms        0.378 ms         2037
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/10000000/10000000/manual_time                           8.39 ms         8.41 ms           86
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/10000000/40000000/manual_time                           16.1 ms         16.1 ms           44
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/10000000/100000000/manual_time                          42.7 ms         42.7 ms           12
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000000/100000000/manual_time                         90.2 ms         90.2 ms            6
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/80000000/240000000/manual_time                           121 ms          121 ms            5
Join<int64_t, int64_t>/left_semi_join_64bit_nulls/50000000/50000000/manual_time                           51.8 ms         51.8 ms           11
Join<int64_t, int64_t>/left_semi_join_64bit_nulls/40000000/120000000/manual_time                          75.4 ms         75.4 ms            7
```
</p>
</details>

File: cpp/src/join/semi_join.cu
Comment (jrhemstad) [inline cpp/src/join/semi_join.cu:56]: Is this a copy/paste remnant or something? There's not hash value being computed here.

File: cpp/src/join/semi_join.cu
Comment (jrhemstad) [inline cpp/src/join/semi_join.cu:139]: Yeah, if you're going to try and do the same optimization as is done in the other joins of using the row hash value as the key and the index as the payload, you're going to need to add the equivalent of `pair_contains` from the `multimap`.

File: cpp/src/join/semi_join.cu
Comment (jrhemstad) [inline cpp/src/join/semi_join.cu:152]: Hm, this actually should just be a `retrieve` and a new algorithm that is the inverse of `retrieve`.

**Comment by vyasr** [comment]
I intended this to be a draft PR, just marked it as such. I pushed this out so that I could get some feedback and we could iron out the gaps with cuco, but this isn't really ready for the big time yet. We'll probably want to wait on further improvements in cuco.

File: cpp/src/join/semi_join.cu
Comment (vyasr) [reply cpp/src/join/semi_join.cu]: Yeah sorry there are a few bits of cruft lying around from early implementations when I copied more of the inner/left/full join logic than is really appropriate for semi/anti joins. Just cleaned that up.

File: cpp/src/join/semi_join.cu
Comment (vyasr) [reply cpp/src/join/semi_join.cu:148]: After our discussions online I think we should punt on adding further functionality to static_map for now and just use this device API until we have a chance to move forward with rapidsai/cuCollections#110.

File: cpp/src/join/semi_join.cu
Comment (vyasr) [reply cpp/src/join/semi_join.cu:135]: Given our discussions I think we can wait on this as well. The same issue exists in other join types and probably isn't worth addressing until NVIDIA/cuCollections#110.

**Comment by vyasr** [comment]
This PR requires NVIDIA/cuCollections#118, but once that's merged I think we can move forward with this largely as is. While there are significant improvements that could be made, they are heavily dependent on refactoring cuCollections and I don't think we benefit too much by trying to implement interim stopgap solutions.

File: cpp/src/join/semi_join.cu
Comment (PointKernel) [inline cpp/src/join/semi_join.cu:52]: I'm a bit confused by the key/value types: this returns a pair of `size_type` and `bool` but the declared types are `hash_value_type` and `size_type`.

File: cpp/src/join/semi_join.cu
Comment (PointKernel) [inline cpp/src/join/semi_join.cu:115]:
```suggestion
    auto const [row_bitmask, _ ] =
```
Looks like `null_count` won't be used.

**Comment by PointKernel** [comment]
This PR depends on https://github.com/NVIDIA/cuCollections/pull/113, otherwise the default hash join allocator won't work here.

File: cpp/src/join/semi_join.cu
Comment (vyasr) [reply cpp/src/join/semi_join.cu:52]: That is a good point. For the current implementation the value is unused and is therefore irrelevant. That matches the old implementation here, but I can change it to insert an arbitrary numeric value instead for now (with a comment). Once we've implemented a set in cuco we should be able to switch to using that, so this is a stopgap solution until then.

**Comment by vyasr** [comment]
> This PR depends on [NVIDIA/cuCollections#113](https://github.com/NVIDIA/cuCollections/pull/113), otherwise the default hash allocator won't work here.

I left a comment to this effect [here](https://github.com/rapidsai/cudf/pull/9666/files#diff-835047ebb75e5848893ff55834f3b834e7482fa258574ff839a305d882d217b9R111), but I agree that we probably shouldn't merge this until that PR is done to avoid unexpected races.

File: cpp/src/join/semi_join.cu
Comment (PointKernel) [reply cpp/src/join/semi_join.cu]: Right, thanks for the clarification.

File: cpp/src/join/semi_join.cu
Comment (PointKernel) [inline cpp/src/join/semi_join.cu:95]: Just realized this PR depends on https://github.com/rapidsai/cudf/pull/9704 as well. We are not passing the input `stream` to `default_allocator` in this way and a [stream allocator adaptor](https://github.com/rapidsai/cudf/pull/9704/files#diff-fe28a4362c6d7a5d462be7c23b9d0a2aab1d8e61cc1a4515f863652c53f01f9fR42) should be used. I will ping Jake to merge the dependencies first.

**Review by PointKernel — CHANGES_REQUESTED** [review]

File: cpp/src/join/semi_join.cu
Comment (vyasr) [reply cpp/src/join/semi_join.cu:95]: Yep and #9704 has another cuco dependency I believe.

Comment (vyasr) [reply cpp/src/join/semi_join.cu]: I think I've adapted the code correctly now, but it's probably worth another quick look from you. Then I think this should be good to go.

**Review by PointKernel — APPROVED** [review]
LGTM

**Comment by PointKernel** [comment]
@vyasr Back to the benchmark results, any idea why the new implementation is slower?

**Comment by vyasr** [comment]
> @vyasr Back to the benchmark results, any idea why the new implementation is slower?

I'm reasonably confident that the performance regression is entirely due to the switch from cudf's concurrent unordered map to the cuco static map, which hasn't benefited from the optimizations you worked on for the multimap. @jrhemstad was fine eating the perf hit for now and postponing optimization because we were trying to get the mixed joins in #9917 up and running ASAP.

However, the work in #9917 shows that the new mixed join code is going to have to be a new kernel rather than a direct adaptation of the existing hash join code because of how we deal with shared memory. Therefore, IMO this PR is no longer a prerequisite for getting mixed joins done for semi/anti joins and that work can happen in parallel, i.e. we could start using cuco's static multimap for mixed joins without merging this PR. @jrhemstad in light of that, do you want to hold off on merging this PR until we've had a chance to do the cuco refactoring and optimized `cuco::static_map`? Then we could avoid a performance degradation in hash semi/anti joins.

**Comment by vyasr** [comment]
rerun tests

**Comment by PointKernel** [comment]
Hmm, `cudf`'s `concurrent_unordered_map` is actually a more naive/unoptimized implementation compared to `cuco::static_map`. Both of them are linear probing while `cuco` is even using the CG-based algorithm by default. I think the test case has relatively low occupancy (or few collisions) which may explain why CG-based algorithms are outperformed. We need a follow-up PR dedicated to detailed profiling and performance optimization.

**Review by codereport — APPROVED** [review]
👍

**Review by jrhemstad — APPROVED** [review]

**Comment by vyasr** [comment]
rerun tests

**Comment by vyasr** [comment]
Discussed offline, we're going to get this merged now and deal with perf later.

@gpucibot merge

**Comment by jlowe** [comment]
> deal with perf later.

Is there an issue to track this?

**Comment by vyasr** [comment]
Just made one in #9973.

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**
